Realtime Cards

valayDave opened the pull request.

valayDave committed:
[default-card][ui] default card realtime updatable
- cleanup the ui component code and add naive progress-bar

seethroughdev commented on metaflow/plugins/cards/ui/src/components/line-chart.svelte:
```diff
@@ -21,32 +21,44 @@
-        legend: {
-          display: false,
+  const createChart = () => {
+    const { config, data, labels } = componentData;
```
Is this better than leaving it reactive above? Since the json loading is async, i'm not entirely sure this is ideal.

valayDave replied:
When i tried runnning it without this code change then the charts were not getting updated. Didn't know why so I tried something like this.

valayDave replied:
In lines that follow, there is a `$` function handling some updates  (Is that incorrectly done ?) I don't know the exact syntax.

seethroughdev replied:
if it works, i would leave as is! 

the next version of svelte will solve for this problem entirely anyway.

seethroughdev reviewed (approved):
Looks good, just want to confirm removing reactivity of those vars is a good idea or not.

seethroughdev commented on metaflow/plugins/cards/card_viewer/viewer.html:
```diff
@@ -0,0 +1,145 @@
+    frame.style.display = 'none';
+    if (cardEl.options.length > 0){
+        const val = cardEl.options[cardEl.selectedIndex].value;
+        frame.src = "/card/" + val;
```
Hey @valayDave , you should be able to fix this by changing it to:

``` javascript
frame.src = `/card/${val}`;

// or

frame.src = "\/card\/" + val;
```

but the first option would be preferred.

valayDave committed:
[default-card][ui] Add custom progress bar

valayDave committed:
[default-card][ui] vegachart component
- Chart support within tables and anywhere across the page
- Chart mutation done based on spec/data changes.

valayDave committed:
[default-card][ui] remove chart.js deps.

valayDave committed:
[card-components] customize updates for Tables/Images
- Set options that allows disabling realtime updates for components that may have lots of data like `Image` / `Table`. This can be useful when we are calling refresh often and don't want to update data which maybe static.

valayDave committed:
[card-components] progressbar / vegacharts

valayDave committed:
[card-components] internal MetaflowCard classes support realtime behavior
- Markdown and Artifact Component made realtime.
- default/blank cards made realtime
- default component ids to realtime-updatable UserComponents

seethroughdev committed:
[card-viewer-html] UI + Structure
- Cleanup HTML template
- Handles edge case where no runs are available
- Shows errors when things are not working

valayDave committed:
[card-components] address comments on Image class
- Image.update supports multi-type object
- remove `disable_updates` from public api

seethroughdev committed:
[default-card][ui] migrate to vite

valayDave committed:
[default-card-python] add runtime check
- Add a variable in the HTML template to enable/disable `metaflow_card_update` function

seethroughdev committed:
[default-card][ui] update progress bar styles

valayDave committed:
[card-server] cli command to expose a card server
- Server helps view realtime updates
- Added a card viewer html file
- Created a simple HTTP based card server that will help showcase the realtime cards from querying the server
- The API response contains the task status for the data api
- reload on run change : A background thread checks if there are new runs created via the latest_run file.
